### PR TITLE
Refs : SLAPI-265 Modify field types for the SDK based on the API documentation

### PR DESCRIPTION
**Pull request by admurray-fb**
# Purpose

This PR modifies the data types of various fields in the sdk model definitions based on the data types defined in the API specs.

# Related Material

See Issue https://github.com/freshbooks/freshbooks-nodejs-sdk/issues/56

### CODE REVIEW

File: packages/api/src/models/Client.ts [hunk 1]
```diff
@@ -27,7 +27,7 @@ export default interface Client {
 	statementToken?: Nullable<string>
 	note?: Nullable<string>
 	mobPhone?: string
-	lastLogin?: Nullable<string>
+	lastLogin?: Nullable<Date>
```
Comment (amcintosh) [inline packages/api/src/models/Client.ts:30]: Probably want to call `transformDateResponse` on this one now.

File: packages/api/src/models/Expense.ts [hunk 2]
```diff
@@ -15,39 +15,39 @@ export enum ExpenseStatus {
...
 	vendor?: Nullable<string>
-	extSystemId?: string
-	staffId: string
+	extSystemId?: number
+	staffId: number
 	date: Date
 	hasReceipt?: boolean
 	accountingSystemId?: string
-	backgroundJobId?: Nullable<string>
+	backgroundJobId?: Nullable<number>
```
Comment (amcintosh) [inline packages/api/src/models/Expense.ts:44]: I'd just take this one out. It's internal, and FBC only. We should just remove it from public API documentation.

File: packages/api/src/models/Invoices.ts [hunk 3]
```diff
@@ -65,60 +65,60 @@ enum InvoiceV3Status {
...
 	paid?: Money
-	parent?: string
+	parent?: number
 	paymentDetails?: string
 	paymentStatus?: PaymentStatus
 	poNumber?: Nullable<string>
 	province?: string
 	returnUri?: Nullable<string>
-	sentId?: string
+	sentId?: string // Couldn't found
```
Comment (amcintosh) [inline packages/api/src/models/Invoices.ts:121]: https://www.freshbooks.com/api/invoices
```
sentid | int | userid of user who sent the invoice, typically 1 for admin
```

File: packages/api/src/models/Invoices.ts [hunk 4]
```diff
@@ -65,60 +65,60 @@ enum InvoiceV3Status {
...
 	language?: string
 	lastOrderStatus?: Nullable<string>
 	lines?: Line[]
 	lName?: string
 	notes?: string
 	organization?: string
 	outstanding?: Money
-	owner?: Owner
-	ownerId?: string
+	owner?: Owner // Not in API - need t o check again.
```
Comment (amcintosh) [inline packages/api/src/models/Invoices.ts:112]: This is returned by `GET /invoices/invoices/<id>` with the include `&include[]=owner`
It returns an object with owner details:

```
userid - INT
fname - STRING
lname - STRING
organization - STRING
email - STRING
```

File: packages/api/src/models/Line.ts [hunk 5]
```diff
@@ -9,12 +9,12 @@ export enum LineType {
 }
 
 export default interface Line {
-	amount?: Money
-	basecampId?: string
+	amount?: Money // There is also another amount as String
+	basecampId?: number
```
Comment (amcintosh) [inline packages/api/src/models/Line.ts:13]: This is internal only.

File: packages/api/src/models/Line.ts [hunk 6]
```diff
@@ -9,12 +9,12 @@ export enum LineType {
...
 	compoundedTax?: boolean
-	date?: Date
+	date?: Date // Not on the API
```
Comment (amcintosh) [inline packages/api/src/models/Line.ts:15]: We should do a transform on this.

File: packages/api/src/models/Business.ts [hunk 7]
```diff
@@ -5,11 +5,11 @@ import BusinessClient, { transformBusinessClientResponse, BusinessClientResponse
 import { Nullable } from './helpers'
 
 export default interface Business {
-	id: string
+	id: number
 	name: string
 	accountId: string
 	address: Nullable<Address>
-	phoneNumber?: Nullable<PhoneNumber>
+	phoneNumber?: string
```
Comment (amcintosh) [inline packages/api/src/models/Business.ts:12]: For me the /business endpoint returns:

```
"phone_number": {
            "id": 67611,
            "phone_number": "18663036061"
        },
```

So I think the object is correct.

File: packages/api/src/models/Expense.ts [hunk 8]
```diff
@@ -15,39 +15,38 @@ export enum ExpenseStatus {
...
 	date: Date
 	hasReceipt?: boolean
 	accountingSystemId?: string
-	backgroundJobId?: Nullable<string>
 	notes?: string
-	extInvoiceId?: string
-	amount: Money
-	expenseId?: string
+	extInvoiceId?: number
+	amount: Money // There are two amounts in the API - one is a string and other an object.
```
Comment (amcintosh) [inline packages/api/src/models/Expense.ts:46]: The string is within the object, is it not?

File: packages/api/src/models/Business.ts (hunk 7 above)
Comment (admurray-fb) [reply packages/api/src/models/Business.ts:12]: OK, that makes sense, here however we have it mentioned as a string - should we update the docs in that case?
https://www.freshbooks.com/api/me_endpoint

File: packages/api/src/models/Invoices.ts [hunk 9]
```diff
@@ -65,43 +65,43 @@ enum InvoiceV3Status {
...
-	id?: string
+	id?: number
 	accountId?: string
 	accountingSystemId?: string
 	address?: string
 	amount?: Money
 	autoBill?: boolean
-	autobillStatus?: Nullable<AutoBillStatus>
-	basecampId?: string
+	autobillStatus?: Nullable<string>
```
Comment (amcintosh) [inline packages/api/src/models/Invoices.ts:74]: Why remove the enum?

File: packages/api/src/models/Invoices.ts [hunk 10]
```diff
@@ -65,43 +65,43 @@ enum InvoiceV3Status {
...
 	country?: string
 	createDate: Date
 	createdAt?: Date
 	currencyCode?: string
 	currentOrganization?: string
-	customerId: string
+	customerId: number
 	datePaid?: Nullable<Date>
-	depositAmount?: Nullable<number>
+	depositAmount?: Nullable<number> // This is an object in the API
```
Comment (amcintosh) [inline packages/api/src/models/Invoices.ts:85]: It should be `Nullable<Money>`

File: packages/api/src/models/Expense.ts (hunk 8 above)
Comment (admurray-fb) [reply packages/api/src/models/Expense.ts:46]: Ahh I see - the code and amount (string) are subfields.

File: packages/api/src/models/Item.ts [hunk 11]
```diff
@@ -44,9 +44,9 @@ function transformItemData({
 		updated: new Date(updated),
 		name,
 		description,
-		qty: Number(qty),
+		qty: !qty ? '' : qty.toString(),
```
Comment (amcintosh) [inline packages/api/src/models/Item.ts:47]: This comes back as a string in the API, so you shouldn't need the logic.

File: packages/api/src/models/Item.ts [hunk 12]
```diff
@@ -44,9 +44,9 @@ function transformItemData({
...
 		sku,
-		inventory: !inventory ? null : Number(inventory),
+		inventory: !inventory ? '' : inventory.toString(),
```
Comment (amcintosh) [inline packages/api/src/models/Item.ts:49]: Same as qty.

File: packages/api/src/models/Line.ts [hunk 13]
```diff
@@ -10,11 +10,11 @@ export enum LineType {
 
 export default interface Line {
 	amount?: Money
-	basecampId?: string
+	basecampId?: number
```
Comment (amcintosh) [inline packages/api/src/models/Line.ts:13]: I think we should just remove this. It's internal-only and deprecated.

File: packages/api/src/models/Line.ts [hunk 14]
```diff
@@ -10,11 +10,11 @@ export enum LineType {
...
 	compoundedTax?: boolean
 	date?: Date
```
Comment (amcintosh) [inline packages/api/src/models/Line.ts:15]: I think we should also take this out. It's not used anymore and not documented https://www.freshbooks.com/api/invoices

File: packages/api/src/models/Invoices.ts (hunk 9 above)
Comment (admurray-fb) [reply packages/api/src/models/Invoices.ts:74]: Adding it back, this is incorrect.

File: packages/api/src/models/Line.ts (hunk 14 above)
Comment (admurray-fb) [reply packages/api/src/models/Line.ts:15]: I notice we have the same for Invoice as well, I am guessing we can take that out as well?

File: packages/api/src/models/Business.ts (hunk 7 above)
Comment (amcintosh) [reply packages/api/src/models/Business.ts]: Sounds good. I'll update.

**Review by amcintosh — APPROVED** [review]

**Pull request merged**